### PR TITLE
wgsl: Add `vectorRange` and `sparseVectorRange` to FPTraits API

### PR DESCRIPTION
**Pull request by zoddicus**
This allows for selection of these generators based on the floating point type, which in-turn allows folding a bunch of case builders together.

Issue #3174

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/shader/execution/expression/call/builtin/normalize.spec.ts [hunk 1]
```diff
@@ -10,47 +10,31 @@ import { makeTestGroup } from '../../../../../../common/framework/test_group.js'
...
+            nonConst ? 'unfiltered' : 'finite',
+            FP[trait].normalizeInterval
+          );
+        },
+      }))
+    )
   )
   .reduce((a, b) => ({ ...a, ...b }), {});
 
 export const d = makeCaseCache('normalize', {
```
Comment (jzm-intel) [inline src/webgpu/shader/execution/expression/call/builtin/normalize.spec.ts:36]: Just 
``` suggestion
export const d = makeCaseCache('normalize', vec_cases);
```
is enough. Same for other cases.

File: src/webgpu/shader/execution/expression/call/builtin/reflect.spec.ts [hunk 2]
```diff
@@ -11,49 +11,33 @@ import { makeTestGroup } from '../../../../../../common/framework/test_group.js'
...
+        },
+      }))
+    )
   )
   .reduce((a, b) => ({ ...a, ...b }), {});
 
 export const d = makeCaseCache('reflect', {
-  ...f32_vec_cases,
-  ...f16_vec_cases,
+  ...vec_cases,
```
Comment (jzm-intel) [inline src/webgpu/shader/execution/expression/call/builtin/reflect.spec.ts:40]: ditto.

File: src/webgpu/shader/execution/expression/call/builtin/refract.spec.ts [hunk 3]
```diff
@@ -90,43 +85,28 @@ function generateCases(
...
+        },
+      }))
+    )
   )
   .reduce((a, b) => ({ ...a, ...b }), {});
 
 export const d = makeCaseCache('refract', {
-  ...f32_vec_cases,
-  ...f16_vec_cases,
+  ...vec_cases,
```
Comment (jzm-intel) [inline src/webgpu/shader/execution/expression/call/builtin/refract.spec.ts:109]: ditto.

File: src/webgpu/shader/execution/expression/call/builtin/mix.spec.ts [hunk 4]
```diff
@@ -18,50 +18,30 @@ import { makeTestGroup } from '../../../../../../common/framework/test_group.js'
...
+    ([2, 3, 4] as const).flatMap(dim =>
+      ([true, false] as const).map(nonConst => ({
+        [`${trait}_vec${dim}_scalar_${nonConst ? 'non_const' : 'const'}`]: () => {
+          return FP[trait].generateVectorPairScalarToVectorComponentWiseCase(
+            FP[trait].sparseVectorRange(dim),
+            FP[trait].sparseVectorRange(dim),
+            // NB: Refactor this when adding scalar ranges to FPTrait API
+            trait === 'f32' ? sparseF32Range() : sparseF16Range(),
+            nonConst ? 'unfiltered' : 'finite',
+            ...FP.f32.mixIntervals
```
Comment (jzm-intel) [inline src/webgpu/shader/execution/expression/call/builtin/mix.spec.ts:40]:
``` suggestion
            ...FP[trait].mixIntervals
```

File: src/webgpu/shader/execution/expression/call/builtin/frexp.spec.ts [hunk 5]
```diff
@@ -111,55 +105,41 @@ function makeVectorCaseExp(v: number | readonly number[], trait: 'f32' | 'f16'):
...
+const fract_vec_cases = (['f32', 'f16'] as const)
+  .flatMap(trait =>
+    ([2, 3, 4] as const).map(dim => ({
+      [`${trait}_vec${dim}_fract`]: () => {
+        return FP[trait].vectorRange(dim).map(v => makeVectorCaseFract(v, trait));
+      },
+    }))
+  )
+  .reduce((a, b) => ({ ...a, ...b }), {});
+
```
Comment (jzm-intel) [inline src/webgpu/shader/execution/expression/call/builtin/frexp.spec.ts:117]:
``` suggestion

// Cases: [f32|f16]_vecN_exp
```

File: src/webgpu/shader/execution/expression/call/builtin/frexp.spec.ts [hunk 6]
```diff
@@ -111,55 +105,41 @@ function makeVectorCaseExp(v: number | readonly number[], trait: 'f32' | 'f16'):
   return { input: toInput(v), expected: toOutput(fs) };
 }
 
+const fract_vec_cases = (['f32', 'f16'] as const)
```
Comment (jzm-intel) [inline src/webgpu/shader/execution/expression/call/builtin/frexp.spec.ts:108]:
``` suggestion
// Cases: [f32|f16]_vecN_fract
const fract_vec_cases = (['f32', 'f16'] as const)
```

File: src/webgpu/shader/execution/expression/call/builtin/faceForward.spec.ts [hunk 7]
```diff
@@ -93,43 +89,27 @@ function generateCases(
...
+        },
+      }))
+    )
   )
   .reduce((a, b) => ({ ...a, ...b }), {});
 
 export const d = makeCaseCache('faceForward', {
-  ...f32_vec_cases,
-  ...f16_vec_cases,
+  ...vec_cases,
```
Comment (jzm-intel) [inline src/webgpu/shader/execution/expression/call/builtin/faceForward.spec.ts:112]: ditto.

File: src/webgpu/shader/execution/expression/call/builtin/faceForward.spec.ts [hunk 8]
```diff
@@ -93,43 +89,27 @@ function generateCases(
...
-      },
-    }))
+// Cases: [f32|f16]_vecN_[non_]const
+const vec_cases = (['f32', 'f16'] as const)
+  .flatMap(trait =>
+    ([2, 3, 4] as const).flatMap(dim =>
+      ([true, false] as const).map(nonConst => ({
+        [`${trait}_vec${dim}_${nonConst ? 'non_const' : 'const'}`]: () => {
+          return generateCases(
+            'f32',
```
Comment (jzm-intel) [inline src/webgpu/shader/execution/expression/call/builtin/faceForward.spec.ts:99]:
``` suggestion
            trait,
```

File: src/webgpu/shader/execution/expression/call/builtin/cross.spec.ts [hunk 9]
```diff
@@ -10,55 +10,38 @@ import { makeTestGroup } from '../../../../../../common/framework/test_group.js'
...
 import { FP } from '../../../../../util/floating_point.js';
-import { sparseVectorF64Range, vectorF16Range, vectorF32Range } from '../../../../../util/math.js';
 import { makeCaseCache } from '../../case_cache.js';
 import { allInputSources, onlyConstInputSource, run } from '../../expression.js';
 
 import { abstractBuiltin, builtin } from './builtin.js';
 
 export const g = makeTestGroup(GPUTest);
 
+const cases = (['f32', 'f16'] as const)
```
Comment (jzm-intel) [inline src/webgpu/shader/execution/expression/call/builtin/cross.spec.ts:20]: Maybe
``` suggestion
// Cases: [f32|f16]_[con_]const
const f32_f16_cases = (['f32', 'f16'] as const)
```

File: src/webgpu/shader/execution/expression/call/builtin/mix.spec.ts [hunk 10]
```diff
@@ -18,50 +18,30 @@ import { makeTestGroup } from '../../../../../../common/framework/test_group.js'
...
-        return FP.f16.generateVectorPairScalarToVectorComponentWiseCase(
-          sparseVectorF16Range(n),
-          sparseVectorF16Range(n),
-          sparseF16Range(),
-          nonConst ? 'unfiltered' : 'finite',
-          ...FP.f16.mixIntervals
-        );
-      },
-    }))
+const vec_scalar_cases = (['f32', 'f16'] as const)
```
Comment (jzm-intel) [inline src/webgpu/shader/execution/expression/call/builtin/mix.spec.ts:29]:
``` suggestion
// [f32|f16]_vecN_scalar_[non_]const
const vec_scalar_cases = (['f32', 'f16'] as const)
```

**Review by jzm-intel — APPROVED** [review]
LGTM with some typo. The change makes cases cleaner, thanks!

File: src/webgpu/shader/execution/expression/call/builtin/normalize.spec.ts [hunk 11]
```diff
@@ -10,47 +10,31 @@ import { makeTestGroup } from '../../../../../../common/framework/test_group.js'
...
-    ([true, false] as const).map(nonConst => ({
-      [`f16_vec${n}_${nonConst ? 'non_const' : 'const'}`]: () => {
-        return FP.f16.generateVectorToVectorCases(
-          vectorF16Range(n),
-          nonConst ? 'unfiltered' : 'finite',
-          FP.f16.normalizeInterval
-        );
-      },
-    }))
+const vec_cases = (['f32', 'f16'] as const)
```
[the remark this answers is not in the document]
Comment (zoddicus) [reply src/webgpu/shader/execution/expression/call/builtin/normalize.spec.ts]: Done

File: src/webgpu/shader/execution/expression/call/builtin/normalize.spec.ts (hunk 1 above)
Comment (zoddicus) [reply src/webgpu/shader/execution/expression/call/builtin/normalize.spec.ts]: Done

File: src/webgpu/shader/execution/expression/call/builtin/cross.spec.ts (hunk 9 above)
Comment (zoddicus) [reply src/webgpu/shader/execution/expression/call/builtin/cross.spec.ts]: Done

File: src/webgpu/shader/execution/expression/call/builtin/faceForward.spec.ts (hunk 8 above)
Comment (zoddicus) [reply src/webgpu/shader/execution/expression/call/builtin/faceForward.spec.ts]: Done

File: src/webgpu/shader/execution/expression/call/builtin/faceForward.spec.ts (hunk 7 above)
Comment (zoddicus) [reply src/webgpu/shader/execution/expression/call/builtin/faceForward.spec.ts]: Done

File: src/webgpu/shader/execution/expression/call/builtin/mix.spec.ts (hunk 10 above)
Comment (zoddicus) [reply src/webgpu/shader/execution/expression/call/builtin/mix.spec.ts:30]: Done

File: src/webgpu/shader/execution/expression/call/builtin/reflect.spec.ts (hunk 2 above)
Comment (zoddicus) [reply src/webgpu/shader/execution/expression/call/builtin/reflect.spec.ts]: Done

File: src/webgpu/shader/execution/expression/call/builtin/refract.spec.ts (hunk 3 above)
Comment (zoddicus) [reply src/webgpu/shader/execution/expression/call/builtin/refract.spec.ts]: Done

File: src/webgpu/shader/execution/expression/call/builtin/frexp.spec.ts (hunk 5 above)
Comment (zoddicus) [reply src/webgpu/shader/execution/expression/call/builtin/frexp.spec.ts]: Done

File: src/webgpu/shader/execution/expression/call/builtin/mix.spec.ts (hunk 4 above)
Comment (zoddicus) [reply src/webgpu/shader/execution/expression/call/builtin/mix.spec.ts]: Done

**Pull request merged**